virtcontainers/interfaces.go

```diff
@@ -65,6 +67,7 @@ type VCSandbox interface {
 	SignalProcess(containerID, processID string, signal syscall.Signal, all bool) error
 	WinsizeProcess(containerID, processID string, height, width uint32) error
 	IOStream(containerID, processID string) (io.WriteCloser, io.Reader, io.Reader, error)
+	PreAddDevices(containerID ContainerConfig) ([]api.Device, error)
```

Comment:
@WeiZhang555 Can you please decouple devices from container here? The original idea is to make storage and devices sandbox level resource, and AddStorage() would be able to add storage to an empty sandbox that has no containers. The AddStorage API need to do the following two things:
1. add storage resource info to sandbox, and save it in persistent storage
2. hotplug the storage devices if necessary

And it does not need to return a Device slice. The storage resources are saved in sandbox and VCSandbox.CreateContainer() can reference existing storage/devices with e.g. the device id.

some pseudo code of `PreAddDevices()`
```
func (s *Sandbox) PreAddStorage(devices []DeviceConfig) error {
        for _, d := range devices {
                s.hypervisor.hotplugDevice(d)
                s.storages = append(s.devices, d)
        }
        s.filesystem.saveDeviceInfo()
}
```

And in VCSandbox.CreateContainer(), we allow containerConfig.Mount to reference an existing device with its ID, and update corresponding container.Mount.BlockDevice to match the referenced device.

Reply:
I agree with you in every point. The difficult point here is even after I did lots of reworks, the devices are still coupled to container in some point.
First one is device is created based on input of `config.DeviceInfo` which contains a `containerPath` in it so it has to bind to a container unless we modify the `DeviceInfo` construct.
Second is every device is store in `sandbox/containerDir/devices.json` and `sandbox/containerDir/mount.json`, we might need to move `sandbox/containerDir/devices.json` to `sandbox/devices.json`. 

I agree with you this is the right way, but I don't think I can finish it before 1.0. That's why I choose another way, keep devices.json under the `containerDir/` as before, and hotplug the devices for container in `PreAddStorage` function, then while we are creating the container, it can find the `device.json` under disk and skip the device hotplug process.

Do you want me to modify to the new way you described? It's a right approach, but there's a big risk that I can't finish it in several days. @bergwolf

Reply:
@WeiZhang555 Please change it so that we do not need to maintain the PreAddStorage API in this form post 1.0.  And frakti won't be able to use it anyway. frakti wants to be able to add storage/device to an empty sandbox. If we cannot make it 1.0, let do post release like 1.1.0.

Reply:
I am gonna try to summarize my understanding here. We want an API that allows to hotplug blindly some storage to the VM without having to relate this to any container semantic.
Given the fact that the caller knows what it's doing, we could have a simple `AddStorage()` invoking some hypervisor function in order to hotplug what is needed as a first step.
As a second step, the caller (frakti most likely) will call later into `CreateContainer()`, providing the appropriate volume description expected for this container. If it wants the container to use the storage previously hotplugged, it will have to describe it through an appropriate `Mount` structure. We could add a specific field `NeedToPlug = false` to this structure to specify nothing needs to be done on the host, but a bind mount, or something else has to happen inside the guest.
Am I missing something or this could simply work ?

Reply:
@bergwolf @WeiZhang555 ^^^

Reply:
@sboeuf You description is correct. One thing to note is that `NeedToPlug` is not necessary in device config and virtcontainers can decide if hotplug is needed depending on the device types.